server/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -354,6 +359,7 @@ public synchronized IndexShard createShard(
                 store,
                 queryCache,
                 mapperService,
+                _ -> translogIndexer.apply(RelationName.fromIndexName(shardId.getIndexName())),
```

Comment:
Shouldn't this use the provided argument? Or can we change the parameter to take a supplier?

Reply:
The parameter is actually only needed for tests, which is annoying.  I think it can be turned into a Supplier once we have reworking IndexShard to remove MapperService, but that's a much bigger change.